[ZEPPELIN-2109][2110] Sortable Helium pkgs & Introduce "INTERPRETER" type

AhyoungRyu opened the pull request:
### What is this PR for?

* Make Helium pkgs sortable
For now, we have 3 types of Helium packages: `VISUALIZATION`, `SPELL` and `APPLICATION`.
`VISUALIZATION` and `SPELL` type of pkgs can be published [NPM registry](https://www.npmjs.com/). Likewise, `APPLICATION` type pkg can be registered in [Maven central repository](http://search.maven.org/). 
But all available Helium packages are not sorted and shown in random order in Helium menu.
To do this, I put "NPM Packages" & "Maven Artifacts" button at top-right corner of Helium menu. 

* Introduce "INTERPRETER" type
As a first step of [ZEPPELIN-1993: Install interpreter from Helium menu](https://issues.apache.org/jira/browse/ZEPPELIN-1993), I added `INTERPRETER` type as a new Helium pkg type. Like [ZEPPELIN-1973](https://issues.apache.org/jira/browse/ZEPPELIN-1973) did, we can retrieve Maven artifact(which has `zeppelin-interpreter` as its dependency) info and save it to Helium online registry as well. 
For more detailed explanation, please see [ZEPPELIN-1993](https://issues.apache.org/jira/browse/ZEPPELIN-1993) and [ZEPPELIN-2110](https://issues.apache.org/jira/browse/ZEPPELIN-2110)'s description.

### What type of PR is it?
Improvement


### What is the Jira issue?
* [ZEPPELIN-2109](https://issues.apache.org/jira/browse/ZEPPELIN-2109) : Make Helium packages sortable in Helium menu
* [ZEPPELIN-2110](https://issues.apache.org/jira/browse/ZEPPELIN-2110) : List community & 3rd party interpreter registered at Maven central repo in Helium menu 

### How should this be tested?
To see how `INTERPRETER` type can be shown in Helium menu,

1. Replace temporarily [ZeppelinConfiguration.java#L46](https://github.com/apache/zeppelin/blob/master/zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java#L46) with [this url](https://raw.githubusercontent.com/AhyoungRyu/spark-notebook-example/master/helium-test.json)

2. Build \w below command and restart
```
 $ mvn clean package -DskipTests -pl 'zeppelin-zengine, zeppelin-server, zeppelin-interpreter'
```

3. Start web dev server under `ZEPPELIN_HOME/zeppelin-web` and browse `localhost:9000`
```
$ yarn run dev:helium
```

4. Go to Helium menu
### Screenshots (if appropriate)
 - How's it look?
![record](https://cloud.githubusercontent.com/assets/10060731/23219279/02fc234e-f962-11e6-86a9-30e7dc0953aa.gif)

 - Package selection by types (`VISUALIZATION`, `SPELL`, `INTERPRETER` and `APPLICATION`)
<img width="500" alt="screen shot 2017-02-23 at 12 51 22 am" src="https://cloud.githubusercontent.com/assets/10060731/23219377/3c020fb4-f962-11e6-849c-6da193414d87.png">


 - Disabled "Enable" button until [ZEPPELIN-1993](https://issues.apache.org/jira/browse/ZEPPELIN-1993) is resolved
<img width="580px" alt="screen shot 2017-02-20 at 2 03 23 am" src="https://cloud.githubusercontent.com/assets/10060731/23104412/c95dccb2-f710-11e6-9602-4159c7182e64.png">


### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

felixcheung commented on zeppelin-web/src/app/helium/helium.controller.js:
```diff
@@ -24,6 +24,28 @@ function HeliumCtrl($scope, $rootScope, $sce, baseUrlSrv, ngToast, heliumService
   $scope.bundleOrder = [];
   $scope.bundleOrderChanged = false;
+  $scope.npmRegistry = 'http://registry.npmjs.org/'
+  $scope.mvnCentralRepository = 'http://repo1.maven.org/maven2/'
```
should these be hardcoded like this? perhaps in some enterprises access could be limited?

AhyoungRyu replied:
@felixcheung Thanks for taking a look. I removed those links since they're not used anymore :)

Leemoonsoo commented:
Thanks @AhyoungRyu for the great contribution.

> I put "NPM Packages" & "Maven Artifacts" button at top-right corner of Helium menu.

My guess is, where package lives is not the most interesting information in end-user might interested. I think package type is more important and useful information to be filtered.
How about having a package type filter instead of repository filter?

i.e. filter by package type : interpreter | visualization | spell | ...

AhyoungRyu commented:
@Leemoonsoo Thanks for your feedback and sorry for my late response. I updated again as you said. 
 - overall looks
![record](https://cloud.githubusercontent.com/assets/10060731/23219544/aed35dc2-f962-11e6-869b-72dead09de96.gif)

 - buttons in header used to sort packages by type
<img width="500" alt="screen shot 2017-02-23 at 12 51 22 am" src="https://cloud.githubusercontent.com/assets/10060731/23219552/b74a0bf4-f962-11e6-8b02-e23f5307c472.png">

AhyoungRyu commented:
Ready for review again

1ambda commented:
I will review and comment soon. Great Feature!

AhyoungRyu commented:
Thanks @1ambda!

1ambda commented on zeppelin-web/src/app/helium/helium.controller.js:
```diff
@@ -24,6 +25,16 @@ function HeliumCtrl($scope, $rootScope, $sce, baseUrlSrv, ngToast, heliumService
+  $scope.numberOfEachPackageByType = {}
+  
+  $scope.allPackageTypes = [HeliumType][0]
+  $scope.pkgListByType = 'VISUALIZATION'
```
Can we display all helium types at first?

AhyoungRyu replied:
Then all interpreters registered in Maven central (about 33 artifacts) also will be shown mixed with all visualization and spell type of packages. I think this is a bit overdoing since currently we can't enable/disable the interpreters in Helium menu. How about showing all packages except `INTERPRETER` type? Not sure which way is better. Please let me know if you have better idea :)

1ambda replied:
@AhyoungRyu  I think we can use pagination if there are a lot of packages.

AhyoungRyu replied:
@1ambda Thanks for your suggestion. I created an Jira ticket for this [ZEPPELIN-2209](https://issues.apache.org/jira/browse/ZEPPELIN-2209). Will start work after this PR merged then :)

1ambda commented:
Tested and works as described.

- LGTM except https://github.com/apache/zeppelin/pull/2037#pullrequestreview-24893904

AhyoungRyu commented:
@1ambda Thanks for the review. Sadly there some conflicts. Let me updated again :)

Leemoonsoo commented on zeppelin-web/src/app/helium/helium.html:
```diff
@@ -44,30 +61,54 @@ <h3 class="new_h3">
+             ng-click="enable(pkgName, pkgSearchResult.pkg.artifact)"
+             ng-if="pkgSearchResult.pkg.type === 'INTERPRETER'"
+             class="btn btn-success disabled btn-xs"
+             style="float:right">Coming soon</div>
```
How about make this button click-able and show dialog that just has instructions for installing interpreter manually using bin/install-interpreter.sh ?

AhyoungRyu commented:
@Leemoonsoo Thanks for your feedback. Addressed your comment like below. Please let me know if we need better guide :)
 - Dialog 
<img width="611" alt="screen shot 2017-03-05 at 4 08 39 am" src="https://cloud.githubusercontent.com/assets/10060731/23581561/7c5429ea-0159-11e7-937f-e64dd4f8b079.png">

 - gif 
![interpreter-type](https://cloud.githubusercontent.com/assets/10060731/23581564/81165228-0159-11e7-9752-1f14a5b066ac.gif)

Leemoonsoo commented:
Looks great. zeppelin-web has build error. Can you take care of it and make CI green?
```
[INFO] $ grunt pre-webpack-dist && webpack && grunt post-webpack-dist 
[INFO] Running "eslint:all" (eslint) task
[INFO] 
[INFO] /Users/moon/Projects/zeppelin-review/pr2037/zeppelin-web/src/app/helium/helium.controller.js
[INFO]   159:14  error  Expected '===' and instead saw '=='  eqeqeq
[INFO] 
[INFO] ✖ 1 problem (1 error, 0 warnings)
```

AhyoungRyu commented:
@Leemoonsoo Thanks for testing it out. CI is green now!

Leemoonsoo commented:
LGTM!

AhyoungRyu commented:
@Leemoonsoo Thanks! 
Will merge into master if there are no more comments on this.
